pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,239 @@
+	//TODO: Implement rate limiting to mitigate any DOS kind of attacks.
+	glog.V(2).Infof("registerPlugin called for socketPath: %s", socketPath)
+	client, conn, err := dial(socketPath)
+	if err != nil {
+		glog.Warningf("Dial failed at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	defer conn.Close()
+	stream, err := client.GetInfo(context.Background())
+	if err != nil {
+		glog.Errorf("Failed to start GetInfo grpc stream at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	defer stream.CloseSend()
+	rqst, err := stream.Recv()
+	if err != nil {
+		glog.Errorf("Failed to receive stream at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	resp := new(watcherapi.RegistrationStatus)
+	if handlerCbkFn, ok := w.handlers[rqst.Type]; ok {
+		var versions []string
+		for _, version := range rqst.SupportedVersions {
+			versions = append(versions, version)
+		}
+		// calls handler callback to verify registration request
+		if verifyErr := handlerCbkFn(rqst.Name, versions, socketPath); verifyErr != nil {
+			resp.PluginRegistered = false
+			resp.Error = fmt.Sprintf("Plugin registration failed with err: %v", verifyErr)
+		} else {
+			resp.PluginRegistered = true
+		}
+	} else {
+		resp.PluginRegistered = false
+		resp.Error = fmt.Sprintf("No handler found registered for plugin type: %s, socket: %s", rqst.Type, socketPath)
+	}
+	glog.V(2).Infof("send register resp: %v", resp)
+	if err := stream.Send(resp); err != nil {
+		glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+		return err
+	}
+
+	for {
+		_, err := stream.Recv()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			glog.Errorf("Recv failed. err: %v", err)
+			return err
+		}
+	}
+
+	if !resp.PluginRegistered {
+		glog.Errorf("Registration failed for plugin type: %s, name: %s, socket: %s, err: %s", rqst.Type, rqst.Name, socketPath, resp.Error)
+		return fmt.Errorf("%s", resp.Error)
+	}
+	glog.Infof("Successfully registered plugin for plugin type: %s, name: %s, socket: %s", rqst.Type, rqst.Name, socketPath)
+	return nil
+}
+
+// Start watches for the creation of plugin sockets at the path
+func (w *Watcher) Start() error {
+	glog.V(2).Infof("Plugin Watcher Start at %s", w.path)
+	w.stopCh = make(chan interface{})
+
+	// Creating the directory to be watched if it doesn't exist yet,
+	// and walks through the directory to discover the existing plugins.
+	if err := w.init(); err != nil {
+		return err
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+
+	err = watcher.Add(w.path)
+	if err != nil {
+		watcher.Close()
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+	w.watcher = watcher
+	startStatusCh := make(chan interface{})
+	go func(watcher *fsnotify.Watcher) {
+		close(startStatusCh)
+		for {
+			select {
+			case event := <-watcher.Events:
+				if event.Op&fsnotify.Create == fsnotify.Create {
```

Comment:
You are not handling plugin removal

Reply:
That would be handled by sub-modules on grpc communication failure.

Reply:
I think there is a real difference between removing the socket and the gRPC connection failing.
One could be assimilated to plugin failure, the other to plugin removal.

Though this could be added later considering the recent PRs and issues regarding that